File: Blind-75/python/arrays-and-hashing/string-encode-and-decode.py
from typing import List
class Solution:

    def encode(self, strs: List[str]) -> str:
        result = ''
        if not strs:
            return result
        for s in strs:
            result += f'{len(s)}#{s}'
        return result
        

    def decode(self, s: str) -> List[str]:
        result = []
        if not s:
            return result
        i = 0
        while i < len(s):
            j = i
            while s[j] != '#':
                j += 1
            ln = int(s[i:j])
            st, en = j+1, j+1+int(ln)
            result.append(s[st:en])
            i = en
        return result

if __name__ == "__main__":
    solution = Solution()
    strs = ["hello", "world"]
    encoded_str = solution.encode(strs)
    print(f"Encoded: {encoded_str}")
    decoded_strs = solution.decode(encoded_str)
    print(f"Decoded: {decoded_strs}")

File: Blind-75/java/arrays-and-hashing/StringEncodeAndDecode.java
import java.util.*;
import java.util.stream.Collectors;

public class StringEncodeAndDecode {
    public static void main(String[] args) {
        List<String> original = List.of("Hello", "World!");
        String encoded = encode(original);
        List<String> decoded = decode(encoded);

        System.out.println("Original: " + original);
        System.out.println("Encoded: " + encoded);
        System.out.println("Decoded: " + decoded);
    }    
    public static String encode(List<String> str) {
        if (str == null || str.isEmpty()){
            return "";
        }
        return str.stream().map(st -> st.length() + "#" + st).collect(Collectors.joining());
    }
    public static List<String> decode(String str) {
        List<String> result = new ArrayList<>();
        if (str == null || str.isEmpty()){
            return result;
        }
        int i = 0;
        while(i < str.length()){
            int j = i;
            while(str.charAt(j) != '#'){
                j++;
            }
            int len = Integer.parseInt(str.substring(i, j));
            result.add(str.substring(j+1, j+1+len));
            i = j+1+len;
        }
        return result;
    }

}
